zcash_primitives/src/transaction/components/sapling.rs

```diff
@@ -1,36 +1,113 @@
+use core::fmt::Debug;
 use ff::PrimeField;
 use group::GroupEncoding;
-
 use std::io::{self, Read, Write};
 
-use zcash_note_encryption::ShieldedOutput;
+use zcash_note_encryption::{ShieldedOutput, COMPACT_NOTE_SIZE};
 
 use crate::{
     consensus,
     sapling::{
         note_encryption::SaplingDomain,
-        redjubjub::{PublicKey, Signature},
+        redjubjub::{self, PublicKey, Signature},
         Nullifier,
     },
 };
 
-use zcash_note_encryption::COMPACT_NOTE_SIZE;
+use super::{amount::Amount, GROTH_PROOF_SIZE};
 
-use super::GROTH_PROOF_SIZE;
+pub type GrothProofBytes = [u8; GROTH_PROOF_SIZE];
 
 pub mod builder;
 
+pub trait Authorization: Debug {
+    type Proof: Clone + Debug;
+    type AuthSig: Clone + Debug;
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Unproven;
+
+impl Authorization for Unproven {
+    type Proof = ();
+    type AuthSig = ();
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct Unauthorized;
+
+impl Authorization for Unauthorized {
+    type Proof = GrothProofBytes;
+    type AuthSig = ();
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct Authorized {
+    pub binding_sig: redjubjub::Signature,
+}
+
+impl Authorization for Authorized {
+    type Proof = GrothProofBytes;
+    type AuthSig = redjubjub::Signature;
+}
+
+#[derive(Debug, Clone)]
+pub struct Bundle<A: Authorization> {
+    pub shielded_spends: Vec<SpendDescription<A>>,
+    pub shielded_outputs: Vec<OutputDescription<A>>,
+    pub value_balance: Amount,
+    pub authorization: A,
+}
+
+impl Bundle<Unauthorized> {
+    pub fn apply_signatures(
+        self,
+        spend_auth_sigs: Vec<Signature>,
+        binding_sig: Signature,
+    ) -> Bundle<Authorized> {
+        assert!(self.shielded_spends.len() == spend_auth_sigs.len());
+        Bundle {
+            shielded_spends: self
+                .shielded_spends
+                .iter()
+                .zip(spend_auth_sigs.iter())
+                .map(|(spend, sig)| spend.apply_signature(*sig))
```

Comment:
These should all be consuming. It's definitely possible to do within the Rust type system (if necessary by defining a temporary):
```suggestion
                .into_iter()
                .zip(spend_auth_sigs.into_iter())
                .map(|(spend, sig)| spend.apply_signature(sig))
```

Reply:
I'm making this irrelevant in https://github.com/zcash/librustzcash/pull/375/commits/a2b6af4230df16838343d34c3a3fdcb6aeb4cef8